src/compiler/checker.ts

```diff
@@ -23553,6 +23581,17 @@ namespace ts {
                         break;
                     case SyntaxKind.CommaToken:
                         return narrowType(type, expr.right, assumeTrue);
+                    // Ordinarily we won't see && and || expressions in control flow analysis because the Binder breaks those

```

Comment:
What about ternary expressions?

Reply:
We narrow within the branches of ternary expressions, but not based on the _result_ of ternary expressions, so I don't think there is anything to do there.